subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/DefaultTaskOutputFilesRepository.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.changedetection.state;
+
+import org.gradle.api.invocation.Gradle;
+import org.gradle.cache.CacheBuilder;
+import org.gradle.cache.CacheRepository;
+import org.gradle.cache.PersistentCache;
+import org.gradle.cache.PersistentIndexedCache;
+import org.gradle.cache.PersistentIndexedCacheParameters;
+import org.gradle.cache.internal.FileLockManager;
+import org.gradle.internal.nativeintegration.filesystem.FileType;
+import org.gradle.util.GradleVersion;
+
+import java.io.Closeable;
+import java.io.File;
+import java.io.IOException;
+import java.util.Collections;
+
+import static org.gradle.cache.internal.filelock.LockOptionsBuilder.mode;
+
+public class DefaultTaskOutputFilesRepository implements TaskOutputFilesRepository, Closeable {
+
+    private final static String CACHE_DISPLAY_NAME = "Build Output Cleanup Cache";
+
+    private final CacheRepository cacheRepository;
+    private final PersistentCache cacheAccess;
+    private final FileSystemMirror fileSystemMirror;
+    private final PersistentIndexedCache<String, Boolean> outputFiles; // The value is true if it is an output file, false if it is a parent of an output file
+
+    public DefaultTaskOutputFilesRepository(CacheRepository cacheRepository, Gradle gradle, FileSystemMirror fileSystemMirror, InMemoryCacheDecoratorFactory inMemoryCacheDecoratorFactory) {
+        this.cacheRepository = cacheRepository;
+        this.cacheAccess = createCache(gradle);
+        this.fileSystemMirror = fileSystemMirror;
+        this.outputFiles = cacheAccess.createCache(cacheParameters(inMemoryCacheDecoratorFactory));
+    }
+
+    private static PersistentIndexedCacheParameters<String, Boolean> cacheParameters(InMemoryCacheDecoratorFactory inMemoryCacheDecoratorFactory) {
+        return new PersistentIndexedCacheParameters<String, Boolean>("outputFiles", String.class, Boolean.class)
+            .cacheDecorator(inMemoryCacheDecoratorFactory.decorator(100000, true));
+    }
+
+    @Override
+    public boolean isGeneratedByGradle(File file) {
+        File absoluteFile = file.getAbsoluteFile();
+        return containsFilesGeneratedByGradle(absoluteFile) || isContainedInAnOutput(absoluteFile);
+    }
```

Comment:
If we want to get a consistent view on the in memory cache, then we would need to. For accessing the cache we don't have to since the in-memory cache decorator takes care of this. We actually cannot lock via `cacheAccess.useCache` since the writing to the backing file system is done in a different thread and would lead to locking problem.
Having said that, I think we are fine without locking since only tasks executing in parallel with the current task can modify outputs which affect the `if` statement and these cannot run in parallel to the current task.